framework/mongo/utils.py

```diff
@@ -61,15 +62,79 @@ def wrapper(cls):
     return wrapper
 
 
-def get_or_http_error(Model, pk, allow_deleted=False):
-    instance = Model.load(pk)
+def get_or_http_error(Model, pk_or_query, allow_deleted=False, display_name=None):
+    """Load an instance of Model by primary key or modularodm.Q query. Raise an appropriate
+    HTTPError if no record is found or if the query fails to find a unique record
+    :param type Model: StoredObject subclass to query
+    :param pk_or_query:
+    :type pk_or_query: either
+      - a <basestring> representation of the record's primary key, e.g. 'abcdef'
+      - a <QueryBase> subclass query to uniquely select a record, e.g.
+        Q('title', 'eq', 'Entitled') & Q('version', 'eq', 1)
+    :param boolean allow_deleted: allow deleted records
+    :param basestring display_name:
+    :return: Model instance
+    """
+    display_name = display_name or ''
+
+    instance = None
+    if isinstance(pk_or_query, QueryBase):
+        try:
+            instance = Model.find_one(pk_or_query)
+        except NoResultsFound:
+            raise HTTPError(http.NOT_FOUND, data=dict(
+                message_long="No {name} record matching that query could be found".format(name=display_name)
+            ))
+        except MultipleResultsFound:
+            raise HTTPError(http.BAD_REQUEST, data=dict(
+                message_long="The query must match exactly one {name} record".format(name=display_name)
+            ))
+    else:
+        instance = Model.load(pk_or_query)
+        if not instance:
+            raise HTTPError(http.NOT_FOUND, data=dict(
+                message_long="No {name} record with that primary key could be found".format(name=display_name)
+            ))
     if not allow_deleted and getattr(instance, 'is_deleted', False):
         raise HTTPError(http.GONE, data=dict(
-            message_long="This resource has been deleted"
-        ))
-    if not instance:
-        raise HTTPError(http.NOT_FOUND, data=dict(
-            message_long="No resource with that primary key could be found"
+            message_long="This {name} record has been deleted".format(name=display_name)
         ))
     else:
         return instance
+
+def autoload(Model, extract_key, inject_key, func):
+    """Decorator to autoload a StoredObject instance by primary key and inject into kwargs. Raises
+    an appropriate HTTPError (see #get_or_http_error)
+
+    :param type Model: StoredObject subclass to query
+    :param basetring extract_key: kwargs key to extract Model instance's primary key
+    :param basestring inject_key: kwargs key to inject loaded Model instance into kwargs
+
+    Example usage: ::
+      def get_node(node_id):
+          node = Node.load(node_id)
+          ...
+
+      becomes
+
+      @autoload(Node, 'node_id', 'node')
+      def get_node(node):
+          ...
+
+    Alternatively:
+      import functools
+      autoload_node = functools.partial(autoload, Node, 'node_id', 'node')
+
+      @autoload_node
```

Comment:
*nods* My confusion has to do with the first usage of the decorator ( `@autoload(Node, 'node_id', 'node')`). A stripped down test case doesn't run when the 4-argument decorator is given only three arguments (or when applied directly), and I don't see it used that way anywhere in the codebase. The docstring says it works- but if that's incorrect, we should remove that example.

Normally when defining a function decorator with arguments, we use a double closure (one for the decorator arguments, and one for the actual function to be decorated)- the `functools.partial` solution is a clever alternative, and that second usage example is confirmed to work.

I'm going to try to keep comments as constructive as possible, but there will be where I may need to ask for clarification before being able to offer a more constructive comment. I'll try to keep moving this along, but with 6k lines am grateful for anything that saves me the time of background investigation.

[Here](https://github.com/abought/osf.io/blob/develop/website/project/decorators.py#L75) is an example of a more conventional function decorator with arguments in the OSF, and below is a sample testcase showing similar syntax to the first docstring example.

```
import functools

def artdeco(a,b,c,func):
    @functools.wraps
    def wrapped(*args, **kwargs):
        print "Proof the decorator arguments are used", a,b,c
        return func(*args, **kwargs)
    return wrapped

@artdeco(1,2,3)
def myfunc(d,e):
    print "Some function arguments", d, e

if __name__ == "__main__":
    myfunc(4,5)

----
File "scratch.py", line 10, in <module>
    @artdeco(1,2,3)
TypeError: artdeco() takes exactly 4 arguments (3 given)

Process finished with exit code 1
```